realm/realm-library/src/main/java/io/realm/internal/NativeObjectReference.java

```diff
@@ -23,43 +23,70 @@
  * This class is used for holding the reference to the native pointers present in NativeObjects.
  * This is required as phantom references cannot access the original objects for this value.
  */
-public final class NativeObjectReference extends PhantomReference<NativeObject> {
+final class NativeObjectReference extends PhantomReference<NativeObject> {
 
-    // Using int here instead of enum to make it faster since the cleanup needs to be called
-    // in a loop to dealloc every native reference.
-    public static final int TYPE_LINK_VIEW = 0;
-    public static final int TYPE_ROW = 1;
+    // Linked list to keep the reference of the PhantomReference
+    private static class ReferencePool {
+        NativeObjectReference head;
+
+        synchronized void add(NativeObjectReference ref) {
+            ref.prev = null;
+            ref.next = head;
+            if (head != null) {
+                head.prev = ref;
+            }
+            head = ref;
+        }
+
+        synchronized void remove(NativeObjectReference ref) {
+            NativeObjectReference next = ref.next;
+            NativeObjectReference prev = ref.prev;
+            ref.next = null;
+            ref.prev = null;
+            if (prev != null) {
+                prev.next = next;
+            } else {
+                head = next;
+            }
+            if (next != null) {
+                next.prev = prev;
+            }
+        }
+    }
 
     // The pointer to the native object to be handled
-    final long nativePointer;
-    final int type;
-    // Use boxed type to avoid box/un-box when access the freeIndexList
-    final Integer refIndex;
+    private final long nativePtr;
+    // The pointer to the native finalize function
+    private final long nativeFinalizerPtr;
+    private final Context context;
+    private NativeObjectReference prev;
+    private NativeObjectReference next;
 
-    NativeObjectReference(int type,
+    private static ReferencePool referencePool = new ReferencePool();
+
+    NativeObjectReference(Context context,
                           NativeObject referent,
-                          ReferenceQueue<? super NativeObject> referenceQueue,
-                          Integer index) {
+                          ReferenceQueue<? super NativeObject> referenceQueue) {
         super(referent, referenceQueue);
-        this.type = type;
-        this.nativePointer = referent.nativePointer;
-        refIndex = index;
+        this.nativePtr = referent.getNativePtr();
+        this.nativeFinalizerPtr = referent.getNativeFinalizerPtr();
+        this.context = context;
+        referencePool.add(this);
     }
 
     /**
      * To dealloc native resources.
      */
     void cleanup() {
```

Comment:
```
public void clear() {
        this.referent = null;
}
```
The `Reference.clear()` implementation is like above which is not necessary since the ref will be GCed after `cleanup()` called. Avoid calling it to make the finalizer daemon loop faster.